src/apps/wall/l7fw.lua

```diff
@@ -0,0 +1,275 @@
+module(..., package.seeall)
+
+-- This module implements a level 7 firewall app that consumes the result
+-- of DPI scanning done by l7spy.
+--
+-- The firewall rules are a table mapping protocol names to either
+--   * a simple action ("drop", "reject", "accept")
+--   * a pfmatch expression
+
+local ffi      = require("ffi")
+local link     = require("core.link")
+local packet   = require("core.packet")
+local datagram = require("lib.protocol.datagram")
+local ether    = require("lib.protocol.ethernet")
+local icmp     = require("lib.protocol.icmp.header")
+local ipv4     = require("lib.protocol.ipv4")
+local ipv6     = require("lib.protocol.ipv6")
+local tcp      = require("lib.protocol.tcp")
+local match    = require("pf.match")
+
+L7Fw = {}
+L7Fw.__index = L7Fw
+
+-- create a new firewall app object given an instance of Scanner
+-- and firewall rules
+function L7Fw:new(config)
+   local obj = { local_ipv4 = config.local_ipv4,
+                 local_ipv6 = config.local_ipv6,
+                 local_macaddr = config.local_macaddr,
+                 scanner = config.scanner,
+                 rules = config.rules,
+                 -- this map tracks flows to compiled pfmatch functions
+                 -- so that we only compile them once per flow
+                 handler_map = {} }
+   return setmetatable(obj, self)
+end
+
+-- called by pfmatch handlers, just drop the packet on the floor
+function L7Fw:drop(pkt, len)
+   packet.free(self.current_packet)
+   return
+end
+
+-- called by pfmatch handler, handle rejection response
+function L7Fw:reject(pkt, len)
+   link.transmit(assert(self.output.reject,
```

Comment:
Yes, I think that makes sense. I put the `assert` there so that having a `reject` link could be optional if that kind of rule isn't used. But it probably makes more sense to make it mandatory in all cases.